share/src/main/java/org/apache/kafka/server/share/fetch/ShareFetch.java

```diff
@@ -0,0 +1,200 @@
+     * @param throwable The error that occurred.
+     */
+    public synchronized void addErroneous(TopicIdPartition topicIdPartition, Throwable throwable) {
+        if (erroneous == null) {
+            erroneous = new HashMap<>();
+        }
+        erroneous.put(topicIdPartition, throwable);
+    }
+
+    /**
+     * Check if the share fetch request is completed.
+     * @return true if the request is completed, false otherwise.
+     */
+    public boolean isCompleted() {
+        return future.isDone();
+    }
+
+    /**
+     * Check if all the partitions in the request have errored.
+     * @return true if all the partitions in the request have errored, false otherwise.
+     */
+    public synchronized boolean isErrored() {
+        return erroneous != null && erroneous.size() == partitionMaxBytes().size();
+    }
+
+    /**
+     * May be complete the share fetch request with the given partition data. If the request is already completed,
+     * this method does nothing. If there are any erroneous partitions, they will be added to the response.
+     *
+     * @param partitionData The partition data to complete the fetch with.
+     */
+    public void maybeComplete(Map<TopicIdPartition, PartitionData> partitionData) {
+        if (isCompleted()) {
+            return;
+        }
+
+        Map<TopicIdPartition, PartitionData> response = new HashMap<>(partitionData);
+        // Add any erroneous partitions to the response.
+        addErroneousToResponse(response);
+        future.complete(response);
+    }
+
+    /**
+     * Maybe complete the share fetch request with the given exception for the topicIdPartitions.
+     * If the request is already completed, this method does nothing. If there are any erroneous partitions,
+     * they will be added to the response.
+     *
+     * @param topicIdPartitions The topic id partitions which errored out.
+     * @param throwable The exception to complete the fetch with.
+     */
+    public void maybeCompleteWithException(Collection<TopicIdPartition> topicIdPartitions, Throwable throwable) {
+        if (isCompleted()) {
+            return;
+        }
+        Map<TopicIdPartition, PartitionData> response = topicIdPartitions.stream().collect(
+            Collectors.toMap(tp -> tp, tp -> new PartitionData()
+                .setErrorCode(Errors.forException(throwable).code())
+                .setErrorMessage(throwable.getMessage())));
+        // Add any erroneous partitions to the response.
+        addErroneousToResponse(response);
+        future.complete(response);
+    }
+
+    /**
+     * Filter out the erroneous partitions from the given set of topicIdPartitions.
+     *
+     * @param topicIdPartitions The topic id partitions to filter.
+     * @return The topic id partitions without the erroneous partitions.
+     */
+    public synchronized Set<TopicIdPartition> filterErroneousTopicPartitions(Set<TopicIdPartition> topicIdPartitions) {
```

Comment:
The input ordering is important. Could we make that explicit?

Reply:
Yes it make sense but the problem is even LinkedHashMap's keySet method returns `Set` interface with an internal implementation of LinkedKeySet. Hence I kept the method definition as `Set` because of the constraint.

Reply:
Could we document that we expect the input to be ordered? Could we change the return type to be LinkedHashSet?

Reply:
Done, added in the method.

As I return the input as it is, when erroneous is null (which should be for most of the requests), without copying hence it should be match i.e. Set.